### PR TITLE
pkg/kvstore: fix etcd segmentation violation

### PR DESCRIPTION
**Pull request by aanm**
Fixes the following panic:

```
goroutine 531 [running]:
go.etcd.io/etcd/clientv3/concurrency.(*Session).Orphan(...)
	/go/src/github.com/cilium/cilium/vendor/go.etcd.io/etcd/clientv3/concurrency/session.go:90
go.etcd.io/etcd/clientv3/concurrency.(*Session).Close(0xc0017b2b70, 0xc003a40570, 0x2)
	/go/src/github.com/cilium/cilium/vendor/go.etcd.io/etcd/clientv3/concurrency/session.go:96 +0x2a
github.com/cilium/cilium/pkg/kvstore.(*etcdClient).Close(0xc001cfe000)
	/go/src/github.com/cilium/cilium/pkg/kvstore/etcd.go:1654 +0x48e
github.com/cilium/cilium/pkg/clustermesh.(*remoteCluster).restartRemoteConnection.func1(0x2b31500, 0xc000fed2c0, 0x3c79c80, 0x0)
	/go/src/github.com/cilium/cilium/pkg/clustermesh/remote_cluster.go:178 +0x40c
github.com/cilium/cilium/pkg/controller.(*Controller).runController(0xc001933200)
	/go/src/github.com/cilium/cilium/pkg/controller/controller.go:205 +0xa77
created by github.com/cilium/cilium/pkg/controller.(*Manager).updateController
	/go/src/github.com/cilium/cilium/pkg/controller/manager.go:120 +0xb50
```

Fixes: c1b05dfba81a ("pkg/kvstore: introduced a dedicated session for locks")
Signed-off-by: André Martins <andre@cilium.io>

```release-note
Fix potential panic on clustermesh environments
```

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
retest-gke

**Review by tklauser — APPROVED** [review]

**Review by christarazi — COMMENTED** [review]
Does this PR fixup https://github.com/cilium/cilium/pull/14623? It seems to be related to the same segfault stacktrace

File: pkg/kvstore/etcd.go [hunk 1]
```diff
@@ -1648,18 +1648,20 @@ func (e *etcdClient) ListPrefix(ctx context.Context, prefix string) (v KeyValueP
 // Close closes the etcd session
 func (e *etcdClient) Close() {
 	close(e.stopStatusChecker)
-	<-e.firstSession
+	sessionErr := e.waitForInitialSession(context.Background())
```
Comment (joestringer) [inline pkg/kvstore/etcd.go:1651]: Why is `context.Background()` OK here? It shouldn't be different from just blocking on a channel, but shouldn't we be passing a real context in?

Comment (aanm) [reply pkg/kvstore/etcd.go:1651]: That's right, it is not different than just blocking on a channel but we need the value of `e.sessionErr()`. The `Close()` does not have a context being passed in.

**Comment by aanm** [comment]
> Does this PR fixup #14623? It seems to be related to the same segfault stacktrace

Yes

**Comment by joestringer** [comment]
I'm fine with merging this with the segfault fix in mind, but unless there's some other timer under the hood, this `context.Background()` makes me nervous that we are just waiting for the time when Cilium can't connect to etcd and gets stuck. I guess at least from this codepath it's in a controller so it shouldn't necessarily block other runtime logic, but that's also part of the point - if this is handled from a controller, then that controller itself has a context so we should be able to pass it down.

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
I filed https://github.com/cilium/cilium/issues/15133 to follow up on my concern above, we can continue the discussion there and decide if further action is necessary.

**Pull request merged**